### PR TITLE
Turn on scope validation for service provider in Development

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/9365

### CODE REVIEW

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs [hunk 1]
```diff
@@ -58,25 +58,20 @@ private WebAssemblyHostBuilder()
 
             InitializeDefaultServices();
 
+            var hostEnvironment = InitializeEnvironment(jsRuntime);
+
             _createServiceProvider = () =>
             {
-                return Services.BuildServiceProvider();
+                return Services.BuildServiceProvider(validateScopes: hostEnvironment.Environment == "Development");
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs]: This is the actual code change. The remainder of this is refactoring to allow unit testing this.

**Review by pranavkm — COMMENTED** [review]
Turning on scope validation does not affect startup - the validation is performed at the call-site i.e. when you request a service. I injected a service in the benchmarking app, which had a fairly deeply nested DI tree and the perf numbers remained unchanged

My deeply nested DI graph:

```
public class Foo1 { public Foo1(Foo2 val, Foo3 val2) {} }
public class Foo2 { public Foo2(Foo3 val, Foo4 val2) {} }
public class Foo3 { public Foo3(Foo4 val, Foo5 val2) {} }
public class Foo4 { public Foo4(Foo5 val, Foo6 val2) {} }
public class Foo5 { public Foo5(Foo6 val, Foo7 val2) {} }
public class Foo6 { public Foo6(Foo7 val, Foo8 val2) {} }
public class Foo7 { public Foo7(Foo8 val, Foo9 val2) {} }
public class Foo8 { public Foo8(Foo9 val, Foo10 val2) {} }
public class Foo9 { public Foo9(Foo10 val, Foo11 val2) {} }
public class Foo10 { public Foo10(Foo11 val, Foo12 val2) {} }
public class Foo11 { public Foo11(Foo12 val, Foo13 val2) {} }
public class Foo12 { public Foo12(Foo13 val, Foo14 val2) {} }
public class Foo13 { public Foo13(Foo14 val, Foo15 val2) {} }
public class Foo14 { public Foo14(Foo15 val, Foo16 val2) {} }
public class Foo15 { public Foo15(Foo16 val, Foo17 val2) {}
```

File: src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs [hunk 2]
```diff
@@ -89,7 +89,7 @@ public TResult InvokeUnmarshalled<TResult>(string identifier)
         /// <param name="arg1">The second argument.</param>
         /// <param name="arg2">The third argument.</param>
         /// <returns>The result of the function invocation.</returns>
-        public TResult InvokeUnmarshalled<T0, T1, T2, TResult>(string identifier, T0 arg0, T1 arg1, T2 arg2)
+        public virtual TResult InvokeUnmarshalled<T0, T1, T2, TResult>(string identifier, T0 arg0, T1 arg1, T2 arg2)
```
Comment (javiercn) [inline src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs]: Is there a way we can avoid this virtual method? If are making this method virtual, shouldn't we make all the overloads virtual for consistency?

**Review by javiercn — APPROVED** [review]
Looks great, I only have a minor comment about the method being virtual.

File: src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs (hunk 2 above)
Comment (pranavkm) [reply src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs]: What do you think about making the other overloads extension methods? They just call in to this overload with default values

**Pull request merged**